lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp

```diff
@@ -0,0 +1,426 @@
+    switch (memportKindIdx) {
+    default:
+      break;
+    case 1:
+      memportKind = MemOp::PortKind::ReadWrite;
+      portLabel = "RW";
+      break;
+    case 2:
+      memportKind = MemOp::PortKind::Write;
+      portLabel = "W";
+      break;
+    }
+
+    // This is set to the count of the kind of memport we're emitting, for
+    // label names.
+    unsigned portNumber = 0;
+
+    // Get an unsigned type with the specified width.
+    auto getType = [&](size_t width) { return UIntType::get(context, width); };
+    auto ui1Type = getType(1);
+    auto addressType = getType(std::max(1U, llvm::Log2_64_Ceil(summary.depth)));
+    auto dataType = UIntType::get(context, summary.dataWidth);
+    auto clockType = ClockType::get(context);
+
+    // Memories return multiple structs, one for each port, which means we
+    // have two layers of type to split apart.
+    for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+      // Process all of one kind before the next.
+      if (memportKind != op.getPortKind(i))
+        continue;
+
+      auto addPort = [&](Direction direction, StringRef field, Type portType) {
+        // Map subfields of the memory port to module ports.
+        auto accesses = getAllFieldAccesses(op.getResult(i), field);
+        for (auto a : accesses)
+          returnHolder[a] = portTypes.size();
+        // Record the new port information.
+        portTypes.push_back(portType);
+        portDirections.push_back(direction);
+        portNames.push_back(
+            builder.getStringAttr(portLabel + Twine(portNumber) + "_" + field));
+      };
+
+      auto getDriver = [&](StringRef field) -> Operation * {
+        auto accesses = getAllFieldAccesses(op.getResult(i), field);
+        for (auto a : accesses) {
+          for (auto *connect : a->getUsers()) {
+            // If this is some use that isn't a connect, move on.
+            if (!isa<ConnectOp, StrictConnectOp>(connect))
+              continue;
+            // If this connect is driving a value to the field, return it.
+            if (connect->getOperand(0) == a)
+              return connect;
+          }
+        }
+        return nullptr;
+      };
+
+      // Find the value connected to the enable and 'and' it with the mask,
+      // and then remove the mask entirely. This is used to remove the mask when
+      // it is 1 bit.
+      auto removeMask = [&](StringRef enable, StringRef mask) {
+        // Get the connect which drives a value to the mask element.
+        auto *maskConnect = getDriver(mask);
+        if (!maskConnect)
+          return;
+        // Get the connect which drives a value to the en element
+        auto *enConnect = getDriver(enable);
+        if (!enConnect)
+          return;
+        // Find the proper place to create the And operation.  The mask and en
+        // signals must both dominate the new operation.
+        OpBuilder b(maskConnect);
+        if (domInfo.dominates(maskConnect, enConnect))
```

Comment:
I made it so that it picks which ever connect between the `en` and `mask` is later in the module and inserts the new connect there.  I guess I could have just inserted the new connect at the bottom of the module and not worried about the order. 
```
mem m

wire en : UInt<1>
m.en <= en ; this connect will be deleted

wire mask : UInt<1>
; insert new connect here
; m.en <= and(mask, end)
m.mask <= mask ; this connect will be deleted
```

The code works by inserting above the `maskConnect` (`OpBuilder b(maskConnect);`).  If the `enConnect` comes after the `maskConnect` then we move the insertion point to above the `enConnect`.  It doesn't matter if we insert right above or right below the old connects because we're going to delete them anyways.